background/rust.tex

```diff
@@ -16,13 +16,64 @@ \subsection{Language Features}
+The example code in \autoref{lst:rust_variable_bindings} shows how to declare both mutable and immutable variables and how to modify their values.
+The example will not work because we try to change the value of an immutable variable.
+
+\begin{listing}[H]
+  \begin{minted}{rust}
+    // bind the value `5' to the immutable variable `a'
+    let a: i32 = 5;
+    // bind the value `10' to the mutable variable `a'
+    let mut b = 10;
+    b = a; // change the value of `b'
+    a = b; // <- error: re-assignment of immutable variable `a'
+  \end{minted}
+  \caption{Variable bindings}
+  \label{lst:rust_variable_bindings}
+\end{listing}
+
+An important part of variable bindings is Rust's way to automatically infer the data type of the variables.
+We can see from the example that \texttt{a} is defined to be a 32-bit integer with the value 5, and because the variable \texttt{b} later gets assigned to \texttt{a}'s value, Rust will automatically infer \texttt{b} to be of the same type.
 
 \subsubsection{Sum types - Enum}
 
-The \textbf{enum} type of Rust is a powerful construct.
-The type is based on sum types and can hold tagged data in addition to named values as enumerated types as found in C.
+A sum type, also known as a tagged union, or an \texttt{enum} in Rust, is a data structure that is used to hold only one out of a small set of possible values.
+Rust's \texttt{enum} construct is a class of \emph{algebraic data types} that are common in functional programming languages, which means that its actual type is formed by combining together other types.
+This makes the \texttt{enum} a powerful feature of the language that can be used to deterministically limit the set of possible outcomes for a type.
+\autoref{lst:rust:option} shows the definition of one of Rust's most commonly used types that is used extensively throughout its standard library and other standalone libraries and applications.
+
+\begin{listing}[H]
+  \begin{minted}{rust}
+    pub enum Option<T> {
+      Some(T),
+      None,
+    }
+  \end{minted}
+  \caption{Definition of Option}
+  \label{lst:rust:option}
+\end{listing}
+
+Every \texttt{Option} variable is an \texttt{enum} that can either have the algebraic value \texttt{Some} or the \emph{named value} \texttt{None}.
+If the variable is is \texttt{Some}, the definition also says that it needs to contain any another type \texttt{T}.
+This is also an example of generics in Rust, \texttt{T} can be anything from an integer to a string.
+The \texttt{Option} type is analogous with a normal C pointer value, which either has the value 0, or \texttt{null} (analogous to \texttt{None}), or a value that that points to a structure at a place in memory (analogous to \texttt{Some}).
+The usage of \texttt{enum} types described in this section become very expressive in combination with \emph{match expressions}, as discussed in \autoref{sub:pattern_matching}.
```

Comment:
I suggest more emphasis on the safety of using `Option`compared to a nullptr. 

Perhaps mentioning that the only safe way of observing the value contained inside the `Option` is to use a `match` expression. And that this guarantees that the variable is `null checked` thus preventing nullpointer dereferences. We should also probably mention the downsides vs a pointer as an `Option` is not a zero cost abstraction.